Plumb SecurityOriginData through the register/unregister Blob URL functions

sysrqb opened the pull request:
#### 5f3df5ab292af0cb9acca6b3dd9fc97ad476dc43
<pre>
Plumb SecurityOriginData through the register/unregister Blob URL functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=250155">https://bugs.webkit.org/show_bug.cgi?id=250155</a>
rdar://103929930

This change plumbs the SecurityOriginData of the main frame down into the Blob
Registry. The SecurityOriginData will be used in a follow-up patch for
partitioning the blob registry. This patch should not make any behavioral
changes.

* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::registerBlobURL):
(WebCore::ThreadableBlobRegistry::unregisterBlobURL):
* Source/WebCore/fileapi/ThreadableBlobRegistry.h: Swap the URL parameter order
because I added it incorrectly in 266021@main

* Source/WebCore/platform/network/BlobRegistry.h:
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::BlobRegistryImpl::registerBlobURL):
(WebCore::BlobRegistryImpl::unregisterBlobURL):
* Source/WebCore/platform/network/BlobRegistryImpl.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::didClose):
(WebKit::NetworkConnectionToWebProcess::registerBlobURL):
(WebKit::NetworkConnectionToWebProcess::unregisterBlobURL):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessPlatformStrategies.cpp:
(WebKit::NetworkProcessPlatformStrategies::createBlobRegistry):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.cpp:
(WebKit::BlobRegistryProxy::registerBlobURL):
(WebKit::BlobRegistryProxy::unregisterBlobURL):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f3df5ab292af0cb9acca6b3dd9fc97ad476dc43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14378 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15032 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16119 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14503 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14767 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16274 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14557 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/15102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/12227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16858 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/12387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/12967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19990 "Build is in progress. Recent messages:Printed configuration") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/13464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/13131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16359 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13683 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11525 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12967 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17306 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13527 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/FormData.cpp:
```diff
@@ -101,10 +101,10 @@ Ref<FormData> FormData::create(bool alwaysStream, Vector<char>&& boundary, Vecto
 }
 
-Ref<FormData> FormData::createMultiPart(const DOMFormData& formData)
+Ref<FormData> FormData::createMultiPart(const DOMFormData& formData, ScriptExecutionContext& context)
```
DOMFormData is a JS type, so it has an associated script execution context. While it doesn't currently subclass ActiveDOMObject or ContextDestructionObserver, I think we should make this change. I propose we make DOMFormData subclass `ContextDestructionObserver`, which should be a trivial change. This should allow us to do less piping of ScriptExecutionContext through, since we'd be able to call domFormData.scriptExecutionContext() to get its context.

cdumez replied:
Note that this could be done as a pre-patch as well to facilitate review.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -4022,6 +4022,7 @@ void FrameLoader::loadDifferentDocumentItem(HistoryItem& item, HistoryItem* from
 
         ResourceRequest requestForOriginalURL(request);
         requestForOriginalURL.setURL(itemOriginalURL);
+        // FIXME When the Blob Registry is partitioned by top-origin, we can't navigate back() from a cross-origin page.
```
`FIXME: ` (missing semicolon)

cdumez commented on Source/WebCore/platform/network/PartitionBlobRegistry.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace WebCore {
+
+enum class PartitionBlobRegistry : bool { No, Yes };
```
We usually like to use prefixes for boolean types to make it clear they're boolean. Here I would suggest `ShouldPartitionBlobRegistry`

cdumez commented on Source/WebCore/Modules/fetch/FetchBody.cpp:
```diff
@@ -250,7 +252,10 @@ RefPtr<FormData> FetchBody::bodyAsFormData() const
         auto body = FormData::create();
-        body->appendBlob(blobBody().url());
+        ASSERT(blobBody().scriptExecutionContext());
+        auto topOrigin = blobBody().scriptExecutionContext() ? blobBody().scriptExecutionContext()->topOrigin().data() : SecurityOrigin::createOpaque()->data();
```
Seems like this should use SecurityOriginData::createOpaque() from your other PR.

cdumez commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1725,7 +1732,15 @@ class CloneSerializer : CloneBase {
+        auto partitionBlobRegistry { PartitionBlobRegistry::No };
+        if (auto* context = executionContext(m_lexicalGlobalObject)) {
+            topOrigin = context->topOrigin().data();
+            partitionBlobRegistry = context->settingsValues().blobRegistryTopOriginPartitioningEnabled ? PartitionBlobRegistry::Yes : PartitionBlobRegistry::No;
```
I think for this particular use case, given that the blob registry is global and in the network process, it would be nice to have a global setting in the network process, rather that pass the setting for each blob registry function call in the WebProcess.

You can still use an experimental feature flag but we'd basically add logic in the UIProcess to deduce the global network process setting for the pages settings. Like, if any page has partitioning enabled, then we should enable partitioning in the network process. Then the WebProcess wouldn't have to worry about the setting at all. I think this should simplify things nicely.

I also discussed it offline with Brady and he seemed onboard with this idea.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
```diff
@@ -120,7 +120,7 @@ BackForwardListItemState toBackForwardListItemState(const WebCore::HistoryItem&
 }
 
-static Ref<FormData> toFormData(const HTTPBody& httpBody)
+static Ref<FormData> toFormData(const HTTPBody& httpBody, String mainFrameURL, WebCore::PartitionBlobRegistry partitionBlobRegistry)
```
`const String&`

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
```diff
@@ -130,14 +130,14 @@ static Ref<FormData> toFormData(const HTTPBody& httpBody)
             formData->appendFileRange(data.filePath, data.fileStart, data.fileLength.value_or(BlobDataItem::toEndOfFile), data.expectedFileModificationTime);
         }, [&] (const String& blobURLString) {
-            formData->appendBlob(URL { blobURLString });
+            formData->appendBlob(URL { blobURLString }, SecurityOrigin::createFromString(mainFrameURL)->data(), partitionBlobRegistry);
```
Since we don't really need a SecurityOrigin, seems we could use `SecurityOriginData::fromURL(URL(mainFrameURL))`

sysrqb committed:
Plumb SecurityOriginData through the register/unregister Blob URL functions
https://bugs.webkit.org/show_bug.cgi?id=250155
rdar://103929930

This change plumbs the SecurityOriginData of the main frame down into the Blob
Registry. The SecurityOriginData will be used in a follow-up patch for
partitioning the blob registry. This patch should not make any behavioral
changes.

* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::registerBlobURL):
(WebCore::ThreadableBlobRegistry::unregisterBlobURL):
* Source/WebCore/fileapi/ThreadableBlobRegistry.h: Swap the URL parameter order
because I added it incorrectly in 266021@main

* Source/WebCore/platform/network/BlobRegistry.h:
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::BlobRegistryImpl::registerBlobURL):
(WebCore::BlobRegistryImpl::unregisterBlobURL):
* Source/WebCore/platform/network/BlobRegistryImpl.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::didClose):
(WebKit::NetworkConnectionToWebProcess::registerBlobURL):
(WebKit::NetworkConnectionToWebProcess::unregisterBlobURL):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessPlatformStrategies.cpp:
(WebKit::NetworkProcessPlatformStrategies::createBlobRegistry):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.cpp:
(WebKit::BlobRegistryProxy::registerBlobURL):
(WebKit::BlobRegistryProxy::unregisterBlobURL):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:
